Csharp pattern matching

BillWagner opened the pull request:
Fixes #1114

BillWagner committed:
outline done

BillWagner committed:
outline done

BillWagner committed:
Finish first draft and samples.

BillWagner committed:
Finish first draft and samples.

BillWagner committed:
remove wrench from Pattern Matching article.

BillWagner committed:
remove wrench from Pattern Matching article.

BillWagner committed:
add section on definitely assigned.

BillWagner committed:
add section on definitely assigned.

BillWagner committed:
proofread

BillWagner committed:
proofread

BillWagner commented:
This is now ready for review. /cc @mairaw @jaredpar @madstorgersen

cartermp commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+
+## Extending to more shape types
+
+As time goes on, you need to support other shape types. As the number
```
"you need" ---> "you may need"

BillWagner replied:
fixed.

cartermp commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+
+As time goes on, you need to support other shape types. As the number
+of conditions you are testing grows, you'll find that using the `is` pattern
+matching expressions becomes cumbersome. In addition to requiring `if`
```
"becomes" ---> "can become"

BillWagner replied:
fixed.

cartermp commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+These rules mean that you are unlikely to accidentally access the result
+of a pattern match expression when that pattern ws not met.
+
+## Extending to more shape types
```
I'd include "switch" in this title somewhere.

BillWagner replied:
fixed.

cartermp commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+of conditions you are testing grows, you'll find that using the `is` pattern
+matching expressions becomes cumbersome. In addition to requiring `if`
+statements on each type you want to check, the `is` expressions are limited
+to testing a type.
```
"testing a type" ---> "testing a single type"

BillWagner replied:
That was intentional: One could use the `is` expression on a base type, which would match on any derived type. I didn't like the wording, but I couldn't come up with anything better that didn't lead down a rabbit hole (a single type or a set of types derived from a single type....)

Do you still think 'single type' is better, or do you have a different suggestion?

cartermp replied:
That's an interesting philosophical question, I suppose.  If I have `if (x is BaseType b)` and pass in a type that derives from `BaseType`, though that expression matches for that entire type hierarchy, I'm still only testing against `BaseType`.  There's no way for me to observe `b` as anything other than a `BaseType`.

I think that given the above, it would make sense to say, "testing if the input matches a single type".

BillWagner replied:
fixed.

cartermp commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+[!code-csharp[ComputeWithSwitchExpression](../../samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs#06_ComputeWithSwitchExpression "Compute with switch expression")]
+
+The pattern matching `switch` statement uses familiar syntax to developers
+that have used the traditional `switch` statement. Each `case` is evaluated
```
Suggested change:

"traditional `switch` statement" ---> "traditional C-style `switch` statement"

BillWagner replied:
fixed.

cartermp commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+are no longer limited to constant values. Removing that limitation means
+that reordering switch sections may change a program's behavior.
+
+When limited to constant values, no more than one switch
```
I found this a bit difficult to parse.  Perhaps it's better to go through the rules for how things are evaluated in a numerical, step-by-step fashion.  You could then follow that up with examples of only constants, only reference types, and mixed/constant reference types.  It may be worth expounding on how the ordering of the cases could affect behavior, and finish up with general guidance.

cartermp commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+for the first switch block. Notice that the statements in this switch block
+do not use either the variables `c` (for the circle), or `s` (for the square).
+Neither of those variables is definitely assigned in this switch block.
+Clearly one has been assigned, but static analysis cannot know which one. For that reason,
```
> Clearly one has been assigned, but static analysis cannot know which one.

Suggested change:

> If either of these cases match, clearly one of the variables has been assigned.  However, it is impossible to tell *which* has been assigned at compile-time, because either case could match at runtime.  For that reason, ...

BillWagner replied:
good suggestion. Fixed.

cartermp commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+introduce a new variable in the `case` statement, or you will only use the
+variable only in the `when` clause.
+
+Having adding those degenarate shapes, let's add a couple more shape types:
```
I don't understand this part of the sentence: "Having adding those degenarate shapes"

BillWagner replied:
updated.

cartermp commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+
+## Conclusions
+
+*Pattern Matching expressions* enable you to easily manage control flow
```
I'd replace "expressions" with "constructs" here, too.

BillWagner replied:
updated.

cartermp reviewed (commented):
Overall I think it's good.  Left some feedback.

mairaw commented on docs/csharp/pattern-matching.md:
````diff
@@ -2,31 +2,192 @@
+```
+
+Language rules for pattern matching expressions help you avoid misuising
+the results of a match expression. In the example above, the variables `s`
````
and what about variable s1? you show it in the code but doesn't talk about it anywhere.

BillWagner replied:
Deleted. (I was checking some syntax and that was left in by mistake.)

mairaw commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+Ideally, you'd want one method that computes the area of any kind of
+shape. You've only created a couple different kinds of shapes so far.
+Let's use an `is` pattern matching expression to test the type of 
+shape:
```
type of `shape` or type of the shape?

BillWagner replied:
fixed.

mairaw commented on samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs:
```diff
@@ -0,0 +1,135 @@
+                return ComputeAreaOfSquare(s);
+            else if (shape is Circle c)
+                return ComputeAreaOfCircle(c);
+//            Console.WriteLine(c);
```
should these comments be removed?

BillWagner replied:
fixed.

mairaw reviewed (changes requested):
Looks good but left a few comments for you to address

BillWagner commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
 
--->
\ No newline at end of file
+The new *pattern matching* expressions enable cleaner syntax to examine data
```
Hmm. I was going from the spec notes here: https://github.com/dotnet/roslyn/blob/features/patterns/docs/features/patterns.md#match-expression  

I'll use 'constructs' here, and read the rest of the doc for the use of both expressions and statements.

BillWagner commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+You'll use *pattern matching* instead. To further emphasize that we're not
+using inheritance, you'll make each shape a `struct` instead of a class. 
+
+## Using `is` pattern expressions ##
```
it ends the header tag. I may have had C# in this heading earlier, and then it's necessary.

BillWagner commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+
+## Using `is` pattern expressions ##
+
+Let's start simple with a square. 
```
fixed.

BillWagner commented on docs/csharp/pattern-matching.md:
````diff
@@ -2,31 +2,192 @@
+} 
+```
+
+Language rules for pattern matching expressions help you avoid misuising
````
fixed.

BillWagner commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+
+Let's examine both of those rules in detail, beginning with scope. The variable
+`c` is in scope only in the `else` branch of the first `if` statement. The variable
+`s` is in scope in the method `ComputeArea`. That's because the true and false
```
I reworded a bit differently. Let me know what you think.

BillWagner commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+Let's examine both of those rules in detail, beginning with scope. The variable
+`c` is in scope only in the `else` branch of the first `if` statement. The variable
+`s` is in scope in the method `ComputeArea`. That's because the true and false
+branches of an `if` statement each establish a scope. However, the `if` statement
```
good point. Added.

BillWagner commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+
+The variables `c` and `s` are assigned when the respective `if` statements are true
+because of the definitely assigned when true mechanism. (You could reverse the
+logic by saying `if !(shape is Square s)` and the variable `s` would be definitely
```
fixed.

bleroy commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+Those techniques aren't possible for data that isn't structured in a class
+hierarchy. In those cases, you'd need to utilize traditional procedural
+imperative constructs like `if` or `switch` statements to manage the control
+flow.
```
Can you add a couple examples of such cases?

BillWagner commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+The variables `c` and `s` are assigned when the respective `if` statements are true
+because of the definitely assigned when true mechanism. (You could reverse the
+logic by saying `if !(shape is Square s)` and the variable `s` would be definitely
+assigned only in the false branch.) 
```
good point. I've updated this section.

bleroy commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+different geometric shapes. But, you'll do it without resorting to object
+oriented techniques and building a class hierarchy for the different shapes.
+You'll use *pattern matching* instead. To further emphasize that we're not
+using inheritance, you'll make each shape a `struct` instead of a class. 
```
The problem I see with that choice is that this exact example is commonly used to show the benefits of OOP: just inherit from a common base abstract shape type, or implement an IShape interface that has an Area property. Specialize the implementation for each shape, and in the end you get code that is clean and elegant. If you're going to keep that example, you should probably explain in more details the benefits of pattern matching when compared to classic OOP.
One way to do this could be to explain that an application may want to factor itself differently, and separate the concern of computing areas from the concern of storing information about shape dimensions. One example I like more for using pattern matching is serialization: nowadays, people prefer to keep serialization logic separate from the serializable types for obvious reasons: there are many ways, present and future, to serialize objects, and you don't want to put additional burden on type authors because of a concern that is not theirs.
Maybe you can keep the shape example while making it clear that it's a toy model, and maybe pointing to other examples where this makes more sense.

bleroy commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+to this method. The code in the method checks the type to see if the parameter
+is either a square or a circle. Notice that the pattern match `is` expression
+can assign a variable as well as test the type. The pattern match expression
+`shape is Square s` is semantically the same as the more verbose syntax:
```
Except the scope of s.

bleroy commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+hierarchy. You can also the control logic to use any condition you test on
+the variable. It enables patterns and idioms that you'll need more often
+as we build more distributed applications, where data and the methods that
+manipulate that data are separate. You'll notice that the shape structs
```
"where data and the methods that manipulate that data are separate": this is very clear, and should probably appear earlier in the article, because it provides the main justification for the feature. The article is heavy on "how" in its present state, but relatively light on "why".

BillWagner replied:
Good point. I moved this up to the intro.

bleroy commented on samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs:
```diff
@@ -0,0 +1,135 @@
+{
+    public static class GeometricUtilities
+    {
+        #region 02_ComputeSquareArea
```
Why use regions?

bleroy commented on samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs:
```diff
@@ -0,0 +1,135 @@
+                case Circle c when c.Radius == 0:
+                    return 0;
+
+                case Square s when s.Side < 5:
```
Why the `when s.Side < 5`?

BillWagner replied:
Removed. It was code I was using to check a few syntax rules.

bleroy commented on samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs:
```diff
@@ -0,0 +1,135 @@
+                case Rectangle r when r.Length == 0 || r.Height == 0:
+                    return 0;
+
+                case Square s when s.Side < 5:
```
Same thing here: why the "when"?

BillWagner replied:
Removed. It was code I was using to check a few syntax rules.

BillWagner commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
 
-<!--
+You are probably familiar with building class hierarchies and creating
+virtual methods and overriden methods to customize object behavior based
```
fixed.

BillWagner commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
-<!--
+You are probably familiar with building class hierarchies and creating
+virtual methods and overriden methods to customize object behavior based
+on the runtime type of the object.
```
fixed

BillWagner commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
--->
\ No newline at end of file
+The new *pattern matching* expressions enable cleaner syntax to examine data
+and manipulate control flow based on any conditon of that data.
```
fixed

BillWagner commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+
+[!code-csharp[CircleDefinition](../../samples/csharp/PatternMatching/PatternMatching/Shapes.cs#03_CircleDefinition "Circle definition")]
+
+We'll also write a method to compute the area of a circle:
```
fixed.

BillWagner commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+
+[!code-csharp[04_ComputeCircleArea](../../samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs#04_ComputeCircleArea "Compute circle area")]
+
+Ideally, you'd want one method that computes the area of any kind of
```
fixed.

BillWagner commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+Language rules for pattern matching expressions help you avoid misuising
+the results of a match expression. In the example above, the variables `s`
+and `c` are only in scope and definitely assigned when the respective
+pattern match expressions have true results. If you try and use either
```
fixed.

BillWagner commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+variable in another location, your code generates compiler errors.
+
+Let's examine both of those rules in detail, beginning with scope. The variable
+`c` is in scope only in the `else` branch of the first `if` statement. The variable
```
the first is correct there.

BillWagner commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+assigned only in the false branch.) 
+
+These rules mean that you are unlikely to accidentally access the result
+of a pattern match expression when that pattern ws not met.
```
fixed.

BillWagner commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
+
+[!code-csharp[RectangleAndTriangle](../../samples/csharp/PatternMatching/PatternMatching/Shapes.cs#08_RectangleAndTriangle "Rectangle and Triangle definition")]
+
+Next, you'll add switch labels and blocks for those shapes in the `ComputeArea` method:
```
fixed.

BillWagner committed:
first review pass.

BillWagner committed:
first review pass.

BillWagner committed:
respond to feedback.

BillWagner committed:
respond to feedback.

mairaw commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,203 @@
 description: Learn about pattern matching expressions in C#
 keywords: .NET, .NET Core, C#
-ms.date: 10/04/2016
+ms.date: 12/09/2016
```
shouldn't this be 01/05/2017?

BillWagner replied:
fixed.

mairaw reviewed (commented):
Changes look good overall

rpetrusha commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,203 @@
+The pattern matching `switch` statement uses familiar syntax to developers
+that have used the traditional C-style `switch` statement. Each `case` is evaluated
+and the code beneath the condition that matches the input variable is
+executed. Code execution cannot "fall through" from one `case` expression
```
I found this confusing and somewhat awkwardly worded. Perhaps "Code execution cannot "fall through" from one `case` expression to the next; the syntax of the `case` statement requires that each `case` end with a `break`, `return`, or `goto`.

BillWagner replied:
done.

rpetrusha commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,203 @@
+transfers to the first `switch` label that matches the `switch` expression.  
+Note that the `default` case will only be executed if no other
+case labels match. The `default` case is evaluated last, regardless
+of its textual order.
```
Perhaps also mention what happens if no conditions are met and there is no default?

BillWagner replied:
done.

rpetrusha commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,203 @@
->
+Pattern matching provides a control flow mechanism that is much more
+generalized than is possible with traditional inheritance hierarchies
+available in Object Oriented practices.
```
I don't completely get the contrast between pattern matching and inheritance -- and then the example uses inheritance, which somehow makes it more confusing. Personally, I would relate this not to inheritance, but to convenience and flexibility -- `is` allows assignment and comparison in a single statement, and `switch` offers a flexibility that C# programmers have wanted for a long time.

BillWagner commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,192 @@
->
-> Learn more about how you can contribute on [GitHub](https://github.com/dotnet/docs/blob/master/CONTRIBUTING.md).
->
+Pattern matching provides a control flow mechanism that is much more
```
I rewrote the intro. Let me know what you think.

BillWagner committed:
respond to feedback

Next is reworking the sample.

BillWagner committed:
respond to feedback

Next is reworking the sample.

BillWagner committed:
Update samples and explanations.

Based on review feedback.
Describe the different patterns better.
Show how the new syntax is an extension of existing statements.
Contrast this code with a OO approach better, and emphasize that one
advantage of this approach is separating data and behavior.

BillWagner committed:
Update samples and explanations.

Based on review feedback.
Describe the different patterns better.
Show how the new syntax is an extension of existing statements.
Contrast this code with a OO approach better, and emphasize that one
advantage of this approach is separating data and behavior.

BillWagner commented:
Updated based on the feedback. Ready for another read @rpetrusha @mairaw @cartermp @bleroy

BillWagner committed:
fix a build error

BillWagner committed:
fix a build error

BillWagner committed:
update publish date.

BillWagner committed:
Merge remote-tracking branch 'origin/csharp-pattern-matching' into csharp-pattern-matching

# Conflicts:
#	docs/csharp/pattern-matching.md

mairaw commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,225 @@
+customize object behavior based on the runtime type of the object.
+
+Those techniques aren't possible for data that isn't structured in a class
+hierarchy. When data and methods are separate you need other tools. The new
```
nit: missing comma after separate

BillWagner replied:
fixed.

mairaw commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,225 @@
+oriented techniques and building a class hierarchy for the different shapes.
+You'll use *pattern matching* instead. To further emphasize that we're not
+using inheritance, you'll make each shape a `struct` instead of a class. 
+As you go though this sample, contrast this code with how it would
```
typo: though

BillWagner replied:
fixed

mairaw commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,225 @@
+As you go though this sample, contrast this code with how it would
+be structured as an object hierarchy. When the data you must
+query and manipulate is not a class hierarchy, pattern matching enables
+very elegent designs.
```
typo: elegent

mairaw commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,225 @@
+for the first `switch` block. Notice that the statements in this `switch` block
+do not use either the variables `c` (for the circle), or `s` (for the square).
+Neither of those variables is definitely assigned in this `switch` block.
+If either of tehse cases match, clearly one of the variables has been assigned.
```
typo: tehse

BillWagner replied:
fixed.

mairaw commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,225 @@
+introduce a new variable in the `case` statement, or you will only use the
+variable only in the `when` clause.
+
+Having adding those shapes with 0 area, let's add a couple more shape types:
```
having adding -> having added?

BillWagner replied:
fixed.

BillWagner replied:
fixed.

BillWagner committed:
respond to feedback.

rpetrusha commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,225 @@
+In this topic, we'll look at the new syntax to show you how it enables
+readable, concise code. Pattern matching enables idioms you'll need more
+often building distributed systems, where data and the methods that manipulate
+that data are separate. 
```
An awkward sentence. Perhaps "Pattern matching enables idioms that you use most often when building..."

BillWagner replied:
updated.

rpetrusha commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,225 @@
+Those techniques aren't possible for data that isn't structured in a class
+hierarchy. When data and methods are separate, you need other tools. The new
+*pattern matching* constructs enable cleaner syntax to examine data
+and manipulate control flow based on any condition of that data.
```
This still isn't completely clear. Can you give a brief, approximately 1 sentence example?

BillWagner replied:
updated.

rpetrusha commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,225 @@
+different geometric shapes. But, you'll do it without resorting to object
+oriented techniques and building a class hierarchy for the different shapes.
+You'll use *pattern matching* instead. To further emphasize that we're not
+using inheritance, you'll make each shape a `struct` instead of a class. 
```
For context, it might be good to add that structs can't explicitly inherit from another type.

BillWagner replied:
Done.

rpetrusha commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,225 @@
+query and manipulate is not a class hierarchy, pattern matching enables
+very elegent designs.
+
+Rather than start with an abstract shape definition, and adding different
```
starting... and no comma after definition

BillWagner replied:
done.

rpetrusha commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,225 @@
+[!code-csharp[ClassicIsExpression](../../samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs#02_ClassicIsExpression "Classic type pattern using is")]
+
+That code above is a classic expression of the *type pattern*: You're testing a variable
+to determine its type, and taking different action based on that type.
```
no comma after type, and "a different action"

BillWagner replied:
done

rpetrusha commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,225 @@
+
+[!code-csharp[IsPatternExpression](../../samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs#03_IsPatternExpression "is pattern expression")]
+
+In this updated version, `is` expression both tests the variable, and assigns
```
the `is` expression... And no comma after variable.

BillWagner replied:
done.

rpetrusha commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,225 @@
+to testing if the input matches a single type.
+
+
+As the number of conditions grows, you'll find that the `switch` pattern
```
This partly repeats a sentence in the previous paragraph. Best to just add this sentence to the end of the last paragraph, and say that "In this case, you'll find..."

BillWagner replied:
done.

rpetrusha commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,225 @@
+[!code-csharp[Switch Type Pattern](../../samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs#05_SwitchTypePattern "Compute with `switch` expression")]
+
+The pattern matching `switch` statement uses familiar syntax to developers
+that have used the traditional C-style `switch` statement. Each `case` is evaluated
```
developers who

BillWagner replied:
done.

rpetrusha commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,225 @@
+## `when` clauses in `case` expressions
+
+You can make special cases for those shapes that have 0 area by using
+a `when` clause on the `case` label. A square with side length of 0, or
```
a side length

BillWagner replied:
fixed.

rpetrusha commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,225 @@
+
+[!code-csharp[ComputeDegenerateShapes](../../samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs#07_ComputeDegenerateShapes "Compute shapes with 0 area")]
+
+This change demonstrates a few important points on the new syntax. First,
```
points of or points about

BillWagner replied:
done.

rpetrusha commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,225 @@
+
+This example introduces two different variables in the two `case` labels
+for the first `switch` block. Notice that the statements in this `switch` block
+do not use either the variables `c` (for the circle), or `s` (for the square).
```
no comma after the closing paren.

BillWagner replied:
fixed.

rpetrusha commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,225 @@
+because either case could match at runtime. For that reason,
+most times when you use multiple `case` labels for the same block, you won't
+introduce a new variable in the `case` statement, or you will only use the
+variable only in the `when` clause.
```
repetition of "only" here makes the sentence difficult to parse.

BillWagner replied:
fixed.

rpetrusha commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,225 @@
+among different variables and types that are not related by an inheritance
+hierarchy. You can also the control logic to use any condition you test on
+the variable. It enables patterns and idioms that you'll need more often
+as we build more distributed applications, where data and the methods that
```
you build...

BillWagner replied:
fixed.

BillWagner committed:
update per feedback

mairaw commented on samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs:
```diff
@@ -0,0 +1,163 @@
+                    return c.Radius * c.Radius * Math.PI;
+                case Rectangle r:
+                    return r.Height * r.Length;
+                //default:
```
did you really want the default commented here?

BillWagner replied:
No. Good catch. Fixed.

mairaw commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,229 @@
+
+This change demonstrates a few important points about the new syntax. First,
+multiple `case` labels can be applied to one `switch` section. The statement
+block is executed when any of those labels is true. In this instance,
```
style true as inline code?

BillWagner replied:
fixed.

mairaw commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,229 @@
+
+*Pattern Matching constructs* enable you to easily manage control flow
+among different variables and types that are not related by an inheritance
+hierarchy. You can also the control logic to use any condition you test on
```
the control logic -> control logic

BillWagner replied:
fixed.

mairaw commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,229 @@
+the variable. It enables patterns and idioms that you'll need more often
+as you build more distributed applications, where data and the methods that
+manipulate that data are separate. You'll notice that the shape structs
+used in this sample do not contain any methods, just read only properties.
```
read only -> read-only

BillWagner replied:
fixed.

mairaw reviewed (changes requested):
found a few more things to fix and this should be good to merge!

BillWagner commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,229 @@
+*pattern matching* constructs enable cleaner syntax to examine data
+and manipulate control flow based on any condition of that data. You already
+write `if` statements and `switch` that test a variable's value. You write `is`
+statements that test a variable's type. *Pattern matching* adds new capabilites
```
fixed

BillWagner commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,229 @@
+As you go through this sample, contrast this code with how it would
+be structured as an object hierarchy. When the data you must
+query and manipulate is not a class hierarchy, pattern matching enables
+very elegent designs.
```
fixed

BillWagner committed:
Respond to feedback.

mairaw commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,229 @@
+
+[!code-csharp[ShapeDefinitions](../../samples/csharp/PatternMatching/PatternMatching/Shapes.cs#01_ShapeDefinitions "Shape definitions")]
+
+From these structures, let' write a method that computes the area
```
nit: let' is missing the s

BillWagner replied:
fixed.

mairaw commented on docs/csharp/pattern-matching.md:
```diff
@@ -2,31 +2,229 @@
+Language rules for pattern matching expressions help you avoid misusing
+the results of a match expression. In the example above, the variables `s`,
+ `c`, and `r` are only in scope and definitely assigned when the respective
+pattern match expressions have `true` results. If you try and use either
```
nit: try and -> try to

BillWagner replied:
fixed.

mairaw reviewed (approved):
Found small nits but other than that let's :shipit: 🎆

BillWagner committed:
Respond to final review.
